File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque

n, m, v = map(int, input().split())

node = [[] for _ in range(n + 1)]
for _ in range(m):  # 방문한 노드 입력
    a, b = (map(int, input().split()))
    node[a].append(b)
    node[b].append(a)
    node[a].sort()  # 노드 정렬  ex) [4, 1] => [1, 4]
    node[b].sort()


def dfs(node, v, visited):
    visited[v] = True
    print(v, end=' ')

    for i in node[v]:
        if not visited[i]:
            dfs(node, i, visited)


def bfs(node, start, visited):
    queue = deque([start])
    visited[start] = True

    while queue:
        s = queue.popleft()
        print(s, end=' ')

        for i in node[s]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True


visited = [False] * (n + 1)  # 방문한 노드 표시
dfs(node, v, visited)
print()
visited = [False] * (n + 1)
bfs(node, v, visited)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Queue;

public class Main {

	private static List<Integer>[] graph;

	public static void main(String[] args) throws Exception {

		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		String[] split = in.readLine().split(" ");

		int n = Integer.parseInt(split[0]);
		int m = Integer.parseInt(split[1]);
		int v = Integer.parseInt(split[2]); // 탐색 시작할 번호

		graph = new ArrayList[n + 1];

		for (int i = 0; i <= n; i++) {
			graph[i] = new ArrayList<Integer>();
		}

		for (int i = 0; i < m; i++) {
			split = in.readLine().split(" ");
			int from = Integer.parseInt(split[0]);
			int to = Integer.parseInt(split[1]);

			graph[from].add(to);
			graph[to].add(from);
		}
		for (List<Integer> g : graph) {
			Collections.sort(g);
		}

		boolean[] visited = new boolean[n + 1];
		dfs(visited, v, n);
		System.out.println();
		bfs(v, n);
	}

	private static void dfs(boolean[] visited, int start, int n) {

		visited[start] = true;
		System.out.print(start + " ");

		for (int next : graph[start]) {
			if (!visited[next]) {
				dfs(visited, next, n);
			}
		}

	}

	private static void bfs(int start, int n) {

		Queue<Integer> queue = new ArrayDeque<>();
		boolean[] visited = new boolean[n + 1];

		queue.offer(start);
		visited[start] = true;

		int curr = 0;
		while (!queue.isEmpty()) {
			curr = queue.poll();
			System.out.print(curr + " ");

			for (int v : graph[curr]) {
				if (!visited[v]) {
					queue.offer(v);
					visited[v] = true;
				}
			}
		}

	}
}